src/display/editor/tools.js

```diff
@@ -889,6 +893,7 @@ class AnnotationEditorUIManager {
       clearTimeout(this.#translationTimeoutId);
       this.#translationTimeoutId = null;
     }
+    this._editorUndoBar?.hide();
```

Comment:
Can't we make a better clean-up ?
The class `EditorUndoBar` should have a method `destroy` to make a full clean-up.

Reply:
Sorry, but what do you mean with "full clean-up" here?

Please keep in mind that the GENERIC viewer supports opening more than one PDF document, so we cannot completely destroy everything related to the undo-bar without things breaking.

Reply:
I was meaning removing the event listeners created in the ctor which can be re-added if the bar is shown again.

Reply:
It seems that https://github.com/mozilla/pdf.js/pull/18900#discussion_r1822997001 wasn't taken into account when implementing this!?

Reply:
Basically, you want something like the following diff which also has the advantage of registering more stuff lazily.
```diff
diff --git a/web/editor_undo_bar.js b/web/editor_undo_bar.js
index 7bb96f65f..36b967640 100644
--- a/web/editor_undo_bar.js
+++ b/web/editor_undo_bar.js
@@ -14,11 +14,13 @@
  */
 
 class EditorUndoBar {
+  #closeButton;
+
   #container;
 
-  #controller = null;
+  #eventBus;
 
-  #instController = null;
+  #initController = null;
 
   isOpen = false;
 
@@ -26,6 +28,8 @@ class EditorUndoBar {
 
   #undoButton;
 
+  #showController = null;
+
   static #l10nMessages = Object.freeze({
     highlight: "pdfjs-editor-undo-bar-message-highlight",
     freetext: "pdfjs-editor-undo-bar-message-freetext",
@@ -38,21 +42,27 @@ class EditorUndoBar {
     this.#container = container;
     this.#message = message;
     this.#undoButton = undoButton;
-
-    this.#instController = new AbortController();
-    const boundHide = this.hide.bind(this);
-    const opts = { signal: this.#instController.signal };
-    closeButton.addEventListener("click", boundHide, opts);
-    eventBus._on("beforeprint", boundHide, opts);
-    eventBus._on("download", boundHide, opts);
+    this.#closeButton = closeButton;
+    this.#eventBus = eventBus;
   }
 
   destroy() {
-    this.#instController.abort();
+    this.#initController?.abort();
+    this.#initController = null;
+
     this.hide();
   }
 
   show(undoAction, messageData) {
+    if (!this.#initController) {
+      this.#initController = new AbortController();
+      const opts = { signal: this.#initController.signal };
+
+      const boundHide = this.hide.bind(this);
+      this.#closeButton.addEventListener("click", boundHide, opts);
+      this.#eventBus._on("beforeprint", boundHide, opts);
+      this.#eventBus._on("download", boundHide, opts);
+    }
     this.hide();
 
     if (typeof messageData === "string") {
@@ -73,7 +83,7 @@ class EditorUndoBar {
     this.isOpen = true;
     this.#container.hidden = false;
 
-    this.#controller = new AbortController();
+    this.#showController = new AbortController();
 
     this.#undoButton.addEventListener(
       "click",
@@ -81,7 +91,7 @@ class EditorUndoBar {
         undoAction();
         this.hide();
       },
-      { signal: this.#controller.signal }
+      { signal: this.#showController.signal }
     );
     this.#undoButton.focus();
   }
@@ -92,8 +102,9 @@ class EditorUndoBar {
     }
     this.isOpen = false;
     this.#container.hidden = true;
-    this.#controller?.abort();
-    this.#controller = null;
+
+    this.#showController?.abort();
+    this.#showController = null;
   }
 }
```

Reply:
@Snuffleupagus done, PTAL